### ISSUE TITLE
PSRAM

### ISSUE THREAD

**Issue by podaen** [bertmelis/espMqttClient#27]
Interested to add support for PSRAM? I am glad to test...

esp32 wrover
Arduino IDE 1.8.16
Arduino ESP 2.0.4

**Comment by bertmelis** [comment]
Which items would you put in psram?
- the queue and packets in psram
- queue in ram and packets in psram
- any other items?

I haven't worked with psram yet but I think it's very doable.

**Comment by bertmelis** [comment]
I'll have a look tomorrow.

Easiest is to have the outgoing packets in psram. Allocation for packets is done on only one line in the code. The queue itself likewise. I have to scan the code to see what else. Buffer might be possible indeed.

Now, I suppose you are fine with a compile time choice to put them in psram.

**Comment by podaen** [comment]
It could be that you have to make additional functions. First detect if there is a size for the PSRAM and then init it. You can choose buffer size differenty.

I use this my shetch.

```
#include "esp_heap_caps.h

const int limit = 10000;
heap_caps_malloc_extmem_enable(limit);
```

**Comment by Pablo2048** [comment]
> The only thing is that PSRAM can not be allocated in global variables. I did a quick look and got my eye on...
> 
> _txBuffer _rxBuffer

That is not correct - the PSRAM is unacessible for DMA so encryption functions (like those, used in SSL) are not working. There is option in SDK config to enable support for non-DMA operations in PSRAM region, but it's experimental IIRC and You need to rebuild the ARduino libraries - see https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/external-ram.html#restrictions

**Comment by podaen** [comment]
For the I2S, things can be initialize as zero DMA. For encryption you are right. I see they are improving there designs to handle more encryption, but nothing right now for that memory I think.

> There is option in SDK config to enable support for non-DMA operations in PSRAM region, but it's experimental IIRC and You need to rebuild the ARduino libraries

I won't do that, you have to do a lot things to get it working and after that, if you start an other project you maybe have to set everything back to it original state.

**Comment by bertmelis** [comment]
I actually don't have any modules with PSRAM so I can't even test atm. I've ordered some though.
But because of this, working with PSRAM is totally new for me and I don't know anything about the limitations.

Ideas and PRs welcome!

**Comment by podaen** [comment]
Hi! It's a basic example, but there are a few things you could do more. Like determing the size automaticly...

```
    if(m_buffer) free(m_buffer);
    m_buffer = NULL;
    if(psramInit() && m_buffSizePSRAM > 0) {
        m_f_psram = true;
        m_buffSize = m_buffSizePSRAM;
        m_buffer = (uint8_t*) ps_calloc(m_buffSize, sizeof(uint8_t));
        m_buffSize = m_buffSizePSRAM - m_resBuffSizePSRAM;
    }
    if(m_buffer == NULL) {
        m_f_psram = false;
        m_buffSize = m_buffSizeRAM;
        m_buffer = (uint8_t*) calloc(m_buffSize, sizeof(uint8_t));
        m_buffSize = m_buffSizeRAM - m_resBuffSizeRAM;
    }
```
After that you can choose what you want to to do with it.

**Comment by bertmelis** [comment]
Sorry, I don't understand what you're trying to say.
This lib allocates most memory as 'global' except for data in the outgoing queue. These outgoing packets are allocated on the heap with a `malloc` call and as I understand are allocated in PSRAM when memory usage requires.

**Comment by podaen** [comment]
I think I saw in the config file that the maximum package size could be set... Unfortualy PSRAM can not be allocated in global variables. But in the member list, it works fine.

> I understand are allocated in PSRAM when memory usage requires.

That's exactly what this code is!

**Comment by bertmelis** [comment]
> I think I saw in the config file that the maximum package size could be set...

Good point. You can set the minimum free memory you need to have to allocate memory for a packet. This however does not take psram into account. I'll take care of that.
